### ISSUE TITLE
Most tests fail on run

### ISSUE THREAD

**Issue by yesilzeytin** [antmicro/verilator-verification-features-tests#456]
Firstly, I recently read your [latest blog post](https://antmicro.com/blog/2023/10/running-simple-uvm-testbenches-in-verilator/) and it looks like great progress, so I appreciate your work here a lot.

I have been trying to run the baseline tests here to check on the UVM support in Verilator's state, yet I couldn't manage to replicate the results here. I have tried using both the [stable](https://github.com/verilator/verilator/tree/stable) and [master](https://github.com/verilator/verilator/tree/master) branches of the official [Verilator](https://github.com/verilator/verilator) repository (I am assuming that [Antmicro Verilator](https://github.com/antmicro/verilator) is not the one since its last update is before the UVM support) to run the tests provided here, yet I observe that only four tests are passing:

![image](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/14f7256a-6481-4a55-a59b-1c230b2c90d3)

![image](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/64718bdf-f742-4026-8ece-320a47610f5b)

Also, here is a screenshot from the **simple_uvm_test** FAIL scenario that looks like a PASS in [your report](https://antmicro.github.io/verilator-verification-features-tests/log.html#s1-s15).

![image](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/75ff9f83-b865-410c-bbc7-5c79574dd3df)

I think that this repository is the answer to an older issue in the [uvm-verilator](https://github.com/antmicro/uvm-verilator) repository, which was [asking about running the UVM tests on Verilator](https://github.com/chipsalliance/uvm-verilator/issues/4) so this issue may be related to that one as well. (I think that the one I should be using is yours and not [Chips Alliance's](https://github.com/chipsalliance/uvm-verilator) since their last update was a significant while before your announcement)

I have also cloned the [uvm-verilator](https://github.com/antmicro/uvm-verilator/tree/2e667ac506c12e91d81bde62fe2c08d992731bd4) repository, but since there were no guidelines, I couldn't figure out the process to use it if it is an additional library. I tried running these tests on a system that has the latest version of Verilator installed and nothing done within this repo besides the commands to run tests, and later tried to copy the [Verilator](https://github.com/verilator/verilator) and [uvm-verilator](https://github.com/antmicro/uvm-verilator) contents to the following folders under this repository hierarchy and run again:

{DIR}\verilator-verification-features-tests\verilator\master
{DIR}\verilator-verification-features-tests\uvm\

![image](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/7877dde0-3d81-4e7a-a158-d9ec92687e99)

Both of these approaches were ineffective. Am I missing out on something? Or else, could there be an issue? And in both cases, having documentation on running UVM on Verilator for these tests would be great to ease the process for people trying to hop in.

**Comment by tgorochowik** [comment]
Hi, thank you for the interest!

The tests currently use mainline Verilator (updated periodically with the use of dependabot) and a custom UVM branch (non-functional changes, required to make Verilator work).

This repository contains submodules, if you use the submodules you are sure to be using the exact same versions that are used to generate the report

At the time of writing this comment it is:
https://github.com/verilator/verilator/tree/feae9ca4aac4e0ede4af718e80aaed460b998936
https://github.com/antmicro/uvm-verilator/tree/2e667ac506c12e91d81bde62fe2c08d992731bd4

Note that some tests (currently only the tests for constrained randomization), use a dedicated Verilator fork (also available as a submodule in this repository). The UVM tests use the stock Verilator though.

If you want to see the exact commands that are used to generate and publish the report, please take a look at the CI configuration file: https://github.com/antmicro/verilator-verification-features-tests/blob/main/.github/workflows/test.yml
You can take a look at the last CI runs to see the logs of this execution, here are the logs of the most recent run (at the time of writing this comment): https://github.com/antmicro/verilator-verification-features-tests/actions/runs/7272076374/job/19813637486
You should be able to replicate the results if you execute the same commands on your machine, you can even use docker to make sure everything is compatible with your setup.
You can also try to use a tool like [act](https://github.com/nektos/act) - it should allow you to directly execute the CI configuration file locally (in docker).

If you follow this, you should be able to replicate the results visible in the published report.

If you want to see a real-world use case for UVM with Verilator, you might be interested in https://github.com/antmicro/verilator-verification-features-tests/pull/421/files - this adds a UVM testbench for one of Veers peripherals, this PR contains all the scripts that are used to execute Verilator with UVM.

**Comment by yesilzeytin** [comment]
Hey, thanks a lot for the detailed response. I just wanted to say that I am currently working on it but have some issues that I'm trying to figure out since I reached an 89/125 success rate, which means that many things are alright but somehow different than yours. However, I want to analyze the situation as detailed as I can before providing feedback, but I wanted to let you know that I'm on it and will post an update soon.

**Comment by yesilzeytin** [comment]
Hi, I did some experiments with getting the environment up and running, and here are my observations:

- With or without act, one shall call
```git submodule update --init```
before doing anything else to be able to run the tests. Otherwise, the internal repo directories are left empty. This step doesn't appear in the README, and the steps above didn't contain it, so I think that its inclusion would be an important step to allow replicating.

- The current **.gitmodules** file seems to have a problematic statement as follows:
![Code_4mBJwi48i8](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/9a3ec235-86a9-4aad-aeae-c8796e264ec1)
This highlighted line is probably fetching from an internal git server, and if one is just trying to clone and run from a computer that is not within that network, they will get a fatal error from calling the command above. Therefore, in my opinion, that line should be updated with the online repository link. I can send a PR if you'd like.

- The current **.gitmodules** file does not include the commit IDs that you mention. Therefore, providing them in a "stable" branch or the "main" branch could be a good idea. I can send a PR related to this again if you'd like.

My current **.gitmodules** file looks like this:
![image](https://github.com/antmicro/verilator-verification-features-tests/assets/29541720/7c37516d-53e6-44b6-ab16-fc02fae28124)

- When using act, I couldn't figure out how to extract the report since it was not generated within the directory. I could only extract terminal logs to a log file using the ">" operator when calling act:
```act > act.log```
So, it would be great if you could also provide some direction regarding these.

- I have managed to pass all the tests that were run as described in your report. However, I don't see "Randomize Constraints" tests being run on my part with act flow. The results are listed below:

> Asserts: 10 tests, 10 passed, 0 failed
> Class-Params: 13 tests, 13 passed, 0 failed
> Class: 5 tests, 5 passed, 0 failed
> Event-Control-Expression: 1 test, 1 passed, 0 failed
> Extra-Uvm-Features: 13 tests, 13 passed, 0 failed
> Rand-Types: 4 tests, 0 passed, 4 failed
> Random-Variables: 5 tests, 0 passed, 5 failed
> Randomize: 2 tests, 0 passed, 2 failed
> Sequences: 6 tests, 0 passed, 6 failed
> Signal-Strengths-Should-Fail: 4 tests, 4 passed, 0 failed
> Signal-Strengths: 21 tests, 21 passed, 0 failed
> Test-Plusargs-Variable: 1 test, 1 passed, 0 failed
> Unpacked-Structs: 2 tests, 2 passed, 0 failed
> Uvm-Testbenches: 1 test, 1 passed, 0 failed
> Uvm-Tests: 19 tests, 19 passed, 0 failed
> Virtual-Interfaces: 2 tests, 2 passed, 0 failed
> Wildcard-Associative: 1 test, 1 passed, 0 failed

When I examine the **.github/workflows/test.yml**, I see that **./gen-tests** is being used. And that file calls from the test files in the **tests** directory. I see the **randomize-constraints** folder there that has the tests within it, as do the other test sets; however, they are somehow not run in the **act** routine. When I examined the log that I extracted from the **act** routine, I got the snippet below:

> [Verilator Verification Features Tests/Test branch master-1               ]   ✅  Success - Main Generate Robot test files
> [Verilator Verification Features Tests/Test branch master-1               ] ⭐ Run Main Run tests
> [Verilator Verification Features Tests/Test branch master-1               ]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/5] user= workdir=
> [Verilator Verification Features Tests/Test branch master-1               ]   | ==============================================================================
> [Verilator Verification Features Tests/Test branch master-1               ]   | Asserts                                                                       
> [Verilator Verification Features Tests/Test branch master-1               ]   | ==============================================================================
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_rose.sv                                                        | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   | make[2]: Leaving directory '/home/unezir/act_uvm_test/verilator-verification-features-tests/verilator/randomize-constraints/src/obj_dbg'
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_stable.sv                                                      | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_named_typed.sv                                                 | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_named_without_parenthesis.sv                                   | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   |       Linking ../../bin/verilator_bin...
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   | g++-11 -static-libgcc -static-libstdc++ -Xlinker -gc-sections -o ../../bin/verilator_bin Verilator.o V3Active.o V3ActiveTop.o V3Assert.o V3AssertPre.o V3Ast.o V3AstNodes.o V3Begin.o V3Branch.o V3Broken.o V3CCtors.o V3CUse.o V3Case.o V3Cast.o V3Cdc.o V3Class.o V3Clean.o V3Clock.o V3Combine.o V3Common.o V3Config.o V3Const__gen.o V3Coverage.o V3CoverageJoin.o V3Dead.o V3Delayed.o V3Depth.o V3DepthBlock.o V3Descope.o V3Dfg.o V3DfgAstToDfg.o V3DfgDfgToAst.o V3DfgOptimizer.o V3DfgPasses.o V3DfgPeephole.o V3DupFinder.o V3Timing.o V3EmitCBase.o V3EmitCConstPool.o V3EmitCFunc.o V3EmitCHeaders.o V3EmitCImp.o V3EmitCInlines.o V3EmitCMain.o V3EmitCMake.o V3EmitCModel.o V3EmitCSyms.o V3EmitMk.o V3EmitV.o V3EmitXml.o V3Error.o V3Expand.o V3File.o V3FileLine.o V3Force.o V3Gate.o V3Global.o V3Graph.o V3GraphAlg.o V3GraphAcyc.o V3GraphPathChecker.o V3GraphTest.o V3Hash.o V3Hasher.o V3HierBlock.o V3Inline.o V3Inst.o V3InstrCount.o V3Life.o V3LifePost.o V3LinkCells.o V3LinkDot.o V3LinkJump.o V3LinkInc.o V3LinkLValue.o V3LinkLevel.o V3LinkParse.o V3LinkResolve.o V3Localize.o V3MergeCond.o V3Name.o V3Number.o V3OptionParser.o V3Options.o V3Order.o V3Os.o V3Param.o V3Partition.o V3PreShell.o V3Premit.o V3ProtectLib.o V3Randomize.o V3Reloop.o V3Sched.o V3SchedAcyclic.o V3SchedPartition.o V3SchedReplicate.o V3SchedTiming.o V3Scope.o V3Scoreboard.o V3Slice.o V3Split.o V3SplitAs.o V3SplitVar.o V3Stats.o V3StatsReport.o V3String.o V3Subst.o V3Table.o V3Task.o V3Trace.o V3TraceDecl.o V3Tristate.o V3TSP.o V3Undriven.o V3Unknown.o V3Unroll.o V3VariableOrder.o V3Waiver.o V3Width.o V3WidthSel.o  V3ParseImp.o V3ParseGrammar.o V3ParseLex.o V3PreProc.o   -lpthread -lm
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_named.sv                                                       | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   | make[2]: Leaving directory '/home/unezir/act_uvm_test/verilator-verification-features-tests/verilator/randomize-constraints/src/obj_opt'
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   | make[1]: Leaving directory '/home/unezir/act_uvm_test/verilator-verification-features-tests/verilator/randomize-constraints/src'
> [Verilator Verification Features Tests/Test branch master-1               ]   | assert_past.sv                                                        | PASS |
> [Verilator Verification Features Tests/Test branch master-1               ]   | ------------------------------------------------------------------------------
> [Verilator Verification Features Tests/Test branch randomize-constraints-2]   ❌  Failure - Main Build Verilator
> [Verilator Verification Features Tests/Test branch randomize-constraints-2] exitcode '2': failure
> [Verilator Verification Features Tests/Test branch randomize-constraints-2] 🏁  Job failed

I see a **❌  Failure - Main Build Verilator** a short while after accessing the **/verilator-verification-features-tests/verilator/randomize-constraints/src** directory, so I suspect that something might be broken within that branch. Do you have any ideas why it could be happening? I tried this on two different WSL Ubuntu machines on different PCs and had the same results. So, I was wondering whether you'd be facing the same issue when running the procedure we discussed above in a fresh clone of this repository.

Thanks again for all the work you did here and your previous guidance, which was hugely helpful. ^^

**Comment by tgorochowik** [comment]
Hi,

Thank you for the details, let me answer your comments:
> * With or without act, one shall call
>   `git submodule update --init`
>   before doing anything else to be able to run the tests. Otherwise, the internal repo directories are left empty. This step doesn't appear in the README, and the steps above didn't contain it, so I think that its inclusion would be an important step to allow replicating.

That's true, it's required, in the CI, this is kind of hidden, but the same is achieved with those lines: https://github.com/antmicro/verilator-verification-features-tests/blob/main/.github/workflows/test.yml#L17-L21

Please feel free to send a PR that adds this to the README, thanks.

> * The current **.gitmodules** file seems to have a SSH-dependent statement as follows:
>   This highlighted line is for an SSH clone if I recall right, and if one is just trying to clone this repository and run from a computer and account that doesn't have a registered SSH key, they will get a fatal error from calling the command above. Therefore, in my opinion, that line should be updated with the online repository link for removing that dependency. I can send a PR if you'd like.

That's true, you need to have an ssh key associated with your github account for this to work. Feel free to send a PR that changes the link to https.

> * The current **.gitmodules** file does not include the commit IDs that you mention. Therefore, providing them in a "stable" branch or the "main" branch could be a good idea. I can send a PR related to this again if you'd like. My current **.gitmodules** file looks like this:

We are using dependabot to periodically update the submodules based on the branch specified in .gitmodules
This needs to stay the way it is now so we keep it reproducible and we get automatic updates from dependabot. 

> * When using act, I couldn't figure out how to extract the report since it was not generated within the directory. I could only extract terminal logs to a log file using the ">" operator when calling act:

That's a good point, I'm not sure if act provides any way to access the output files in an easy way, it's probably worth checking their documentation.

Alternatively you could leverage the fact that act uses docker underneath - you should be able to interactively connect to the docker container after act finishes, and then copy the files to your file system.

> I see a **❌  Failure - Main Build Verilator** a short while after accessing the **/verilator-verification-features-tests/verilator/randomize-constraints/src** directory, so I suspect that something might be broken within that branch. Do you have any ideas why it could be happening? I tried this on two different WSL Ubuntu machines on different PCs and had the same results. So, I was wondering whether you'd be facing the same issue when running the procedure we discussed above in a fresh clone of this repository.

Sounds like it cannot access the Verilator fork we use for randomized constraints. Please make sure the submodule is correctly fetched.
BTW. This submodule also contains some submodules of it's own, so instead of the command you suggested at the beginning, please try using:
```
git submodule update --init --recursive
```
This should fetch both the submodules here, but also the submodules of those submodules.

**Comment by yesilzeytin** [comment]
Hi, thanks for all the responses. I'll make pull requests as soon as I can. I later found out that the repo uses **.gitmodules** when running **git submodules update --init**. However, it uses **.config** when **git submodules update --init --recursive** is used. So, I'll send a PR regarding it too.

On the randomize-constraints build, you were right. Apparently, the submodules of the submodules weren't being set up. However, when I did the routine you asked, during the [crave](https://github.com/antmicro/crave/tree/ad6803c9265e3d9295955075eda43e78b92a47ab) build, which is a submodule of the [randomize-constraints branch](https://github.com/antmicro/verilator-1/tree/640b3b9e476770d83cc8a275ea741311356d36f0) I got a build error:

> Now type 'make' (or sometimes 'gmake') to build Verilator.
> pod2man bin/verilator verilator.1
> pod2man bin/verilator_coverage verilator_coverage.1
> cmake -S crave -B crave/build -DmetaSMT_USE_STP=ON -DCMAKE_INSTALL_PREFIX=.
> make: cmake: No such file or directory
> make: *** [Makefile:505: crave/build] Error 127
> make: *** Waiting for unfinished jobs....

I made sure that the most recent **cmake** was installed and also created a **build** folder within the **crave** directory, which didn't exist. However, the issue persisted, and I am not knowledgeable about it. I tried installing both the specific **crave** build provided and the main branch of the **crave**. I got all the builds completed without any error messages, and only tests of the repo failed with the message:

> ctest --test-dir build -j4
> Internal ctest changing into directory: /home/unezir/act_uvm_test/crave/build
> Test project /home/unezir/act_uvm_test/crave/build
> No tests were foound!!!

Since the **crave** repo claims that documentations are outdated, I guess that tests not being able to run may be because of that, and I may open an issue regarding it on that repository. However, I got all the builds completed successfully when I tried standalone installation. Therefore, I think that if I am getting a build error using the **act** routine, that is likely because of some problem within the workflow. However, I couldn't pinpoint the issue there since I'm not knowledgeable about **crave** and also documentation is lacking on that part for me to learn as well. So, I suspect that there might be an issue regarding **crave** build workflow. I am not totally sure, but there are build commands for verilator within the **test.yml** file, while there are no arguments for building **crave**.

I later built everything manually without **act** flow, including the **crave**. When I run tests like that, I observe that **randomize-constraints** tests are being run, but they all fail. I observe that failure is crave-related as well, since the error logs within the generated report explicitly state the following statement:

**constraint_set.sv:**
> In file included from Vtop___024root.h:9,
>                  from Vtop__Syms.h:17,
>                  from Vtop.cpp:5,
>                  from Vtop__ALL.cpp:3:
> /home/unezir/manual_uvm_test/verilator-verification-features-tests/./verilator/randomize-constraints/include/verilated_random.h:15:10: fatal error: crave/ConstrainedRandom.hpp: No such file or directory
>    15 | #include <crave/ConstrainedRandom.hpp>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.

It seems like even though **crave** build gives no errors after running the following commands:

```git clone --recursive [<URL>](https://github.com/antmicro/crave.git)
cmake -B build -DmetaSMT_USECVC4=ON
cmake --build build -j4
cmake -B build -DmetaSMT_USECVC4=ON -DCRAVE_ENABLE_TESTS=ON```

The build fails, and **verilated_random.h** doesn't recognize **crave**. Is it possible if you are running a different routine for building from the one provided in your **crave** repository?

**Comment by tgorochowik** [comment]
You can see the build used to generate the report in the CI: https://github.com/antmicro/verilator-verification-features-tests/actions/runs/7354565734/job/20021953524 (this is a build from Today).

As you can see in the CI definition: https://github.com/antmicro/verilator-verification-features-tests/blob/main/.github/workflows/test.yml#L71 
the commands used to build Verilator are the same for both vanilla Verilator and Verilator with crave.

Can you replicate the build in act? (assuming the submodules are correctly fetched).

**Comment by yesilzeytin** [comment]
Hello again! Sorry for the late response, holiday got in-between before I could make sure everything is alright. I managed to get the builds up and running through the responses you provided. Thanks a lot for all the assistance you provided so far. Now I get 10 tests passing out of 19 with the most recent branches of the submodules. I think that numbers vary with the versions used, so if you don't particularly think there is an issue with that, we can close the issue.

I also created a PR regarding the topics we talked, and added in some additional utilities that I hope to be useful: [Workflow improvement for ease of replicability](https://github.com/antmicro/verilator-verification-features-tests/pull/466)

I hope you have a great 2024!
